ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,358 @@
+      for (CycleLog cycleLog : cycleLogs) {
+        stream.write(cycleLog.toBytes());
+      }
+      stream.writeLong(crcOutputStream.getValue());
+      fileOutputStream.getChannel().force(true);
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+    if (!tempFile.renameTo(file)) {
+      throw new IllegalStateException("Newly written compaction log could not be saved");
+    }
+  }
+
+  /**
+   * Details and log for a single compaction cycle.
+   */
+  private static class CycleLog {
+    private static final int TIMESTAMP_SIZE = 8;
+
+    // details about the cycle
+    final CompactionDetails compactionDetails;
+    // start time of the copy phase
+    long copyStartTime = -1;
+    // start time of the switch phase
+    long switchStartTime = -1;
+    // start time of the cleanup phase
+    long cleanupStartTime = -1;
+    // end time of the current cycle
+    long cycleEndTime = -1;
+    // point until which copying is complete
+    StoreFindToken safeToken = new StoreFindToken();
+
+    /**
+     * Create a log for a single cycle of compaction.
+     * @param compactionDetails the details for the compaction cycle.
+     */
+    CycleLog(CompactionDetails compactionDetails) {
+      this.compactionDetails = compactionDetails;
+    }
+
+    /**
+     * Create a log for a compaction cycle from a {@code stream}.
+     * @param stream the {@link DataInputStream} that represents the serialized object.
+     * @param storeKeyFactory the {@link StoreKeyFactory} used to generate the {@link StoreFindToken}.
+     * @return a {@link CycleLog} that represents a cycle.
+     * @throws IOException if there is an I/O error while reading the stream.
+     */
+    static CycleLog fromBytes(DataInputStream stream, StoreKeyFactory storeKeyFactory) throws IOException {
+      CompactionDetails compactionDetails = CompactionDetails.fromBytes(stream);
+      CycleLog cycleLog = new CycleLog(compactionDetails);
+      cycleLog.copyStartTime = stream.readLong();
+      cycleLog.switchStartTime = stream.readLong();
+      cycleLog.cleanupStartTime = stream.readLong();
+      cycleLog.cycleEndTime = stream.readLong();
+      cycleLog.safeToken = StoreFindToken.fromBytes(stream, storeKeyFactory);
+      return cycleLog;
+    }
+
+    /**
+     * @return the current phase of this cycle of compaction.
+     */
+    Phase getPhase() {
+      Phase phase;
+      if (copyStartTime == -1) {
+        phase = Phase.PREPARE;
+      } else if (switchStartTime == -1) {
+        phase = Phase.COPY;
+      } else if (cleanupStartTime == -1) {
+        phase = Phase.SWITCH;
+      } else if (cycleEndTime == -1) {
+        phase = Phase.CLEANUP;
+      } else {
+        phase = Phase.DONE;
+      }
+      return phase;
+    }
+
+    /**
+     * @return serialized version of the {@link CycleLog}.
+     */
+    byte[] toBytes() {
+      /*
+        Description of serialized format
+
+        compactionDetails (see CompactionDetails#toBytes())
```

Comment:
any reason to serialize the compactionDetails first and then other instance variables? Usually I have seen any such lists will be serailized in the end (for eg, list of cycle logs in CompactionLog).